bin/oay/src/services/webdav/webdavfs.rs

```diff
@@ -48,11 +52,27 @@ impl DavFileSystem for WebdavFs {
 
     fn read_dir<'a>(
         &'a self,
-        _path: &'a dav_server::davpath::DavPath,
+        path: &'a dav_server::davpath::DavPath,
         _meta: dav_server::fs::ReadDirMeta,
     ) -> dav_server::fs::FsFuture<dav_server::fs::FsStream<Box<dyn dav_server::fs::DavDirEntry>>>
     {
-        todo!()
+        async move {
+            let path = path.as_url_string();
+            let mut lister = self.op.list(path.as_str()).await.unwrap();
```

Comment:
Code here seems incorrect. List will list the path from start everytime, and `next_page` only list one page in this dir.

Reply:
so here should loop `next_page` untill `has_next` to be false ? BTW, how many entry in one page ?

Reply:
> so here should loop `next_page` untill `has_next` to be false ?

`Lister` itself is a `Stream`, you can use `while Ok(entry) in lister.next().await {}` here.


> BTW, how many entry in one page ?

These are service implementation details that users should not be aware of.

Reply:
> BTW, how many entry in one page ?
These are service implementation details that users should not be aware of.

just my own curiosity hhh

Reply:
The code logic here will collect all items into vec and than transformed into stream again. How about implement `dav_server::fs::FsStream` directly? We can add a newtype like `DevStream(Lister)` and `impl FsStream for DevStream`.

Reply:
`FsStream` is a type, not a trait 🤔

Reply:
> `FsStream` is a type, not a trait thinking

It's an aliase to:

```rust
pub type FsStream<T> = Pin<Box<dyn Stream<Item = T> + Send>>;
```

We just need to return a `Pin<Box<dyn Stream<Item = T>>>`

Reply:
My idea is:
```
fn read_dir<'a>(
        &'a self,
        path: &'a dav_server::davpath::DavPath,
        _meta: dav_server::fs::ReadDirMeta,
    ) -> dav_server::fs::FsFuture<dav_server::fs::FsStream<Box<dyn dav_server::fs::DavDirEntry>>>
{
  async move {
      let path = path.as_url_string();
      let mut lister = self.op.list(path.as_str()).await.unwrap();
      let entry_stream = DevStream(lister).boxed();
      Ok(entry_stream)
  }
  .boxed()
}


struct DevStream(Lister);

impl Stream for DevStream {
    type Item = Box<dyn DavDirEntry>;

    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Option<Self::Item>> {
       // use `Lister` poll_next()
    }
}
```
But the problem is `Item` in `Lister` is `Result<Entry>`, we need to make it to a `WebDAVDirEntry`, which need a `op` to construct, but we cannot get `op` in `DevStream::poll_next()`. 

Or, can we make a `Metadata` from a `Entry` without `op`? It's required in `WebDAVDirEntry::metadata()`. Any suggestion?

Reply:
We can have a `DevStream { op: Operator, lister: Lister }`